pkg/controller/status/status.go

```diff
@@ -275,6 +279,62 @@ func (m *statusManager) AddCertificateSigningRequests(name string, labels map[st
 	m.certificatestatusrequests[name] = labels
 }
 
+type windowsNodeUpgrades struct {
+	nodesPending    map[string]struct{}
+	nodesInProgress map[string]struct{}
+	nodesCompleted  map[string]struct{}
+}
+
+func newWindowsNodeUpgrades() *windowsNodeUpgrades {
+	return &windowsNodeUpgrades{
+		nodesPending:    make(map[string]struct{}),
+		nodesInProgress: make(map[string]struct{}),
+		nodesCompleted:  make(map[string]struct{}),
+	}
+}
+
+func (w *windowsNodeUpgrades) progressingReason() string {
+	inProgress := len(w.nodesInProgress)
+	pending := len(w.nodesPending)
+	completed := len(w.nodesCompleted)
+	total := pending + inProgress + completed
+
+	if pending+inProgress == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("Waiting for Calico for Windows to be upgraded: %v/%v nodes have been upgraded, %v in-progress", completed, total, inProgress)
+}
+
+func (w *windowsNodeUpgrades) updateNodeUpgrade(node string, inProgress bool) {
+	if inProgress {
+		delete(w.nodesPending, node)
+		w.nodesInProgress[node] = struct{}{}
+	} else {
+		w.nodesPending[node] = struct{}{}
+	}
+}
+
+func (w *windowsNodeUpgrades) removeNodeUpgrade(nodeName string) {
+	delete(w.nodesPending, nodeName)
+	delete(w.nodesInProgress, nodeName)
+	w.nodesCompleted[nodeName] = struct{}{}
+
+	// If there are no more nodes upgrading or pending to be upgraded then clear
+	// the completed map.
+	if len(w.nodesPending) == 0 && len(w.nodesInProgress) == 0 {
+		w.nodesCompleted = make(map[string]struct{})
+	}
+}
+
+// AddWindowsNodeUpgrade tells the status manager to monitor the health of the given
+// Windows node upgrade.
+func (m *statusManager) AddWindowsNodeUpgrade(nodeName string, inProgress bool) {
```

Comment:
This API is confusing me a little bit. I think we could probably simplify it to a single call:

```go 
func SetWindowsUpgradeStatus(pending, inProgress, completed []string) {
}
```

We can do this because the calling code re-calculates these on every iteration anyway.

The main issues I see with the current approach is that it's unclear what `inProgress` means relative to `Add` and `Remove`, and that the status manager needs to attempt to reconstruct the caches that the controller is already building, which means they might get out of sync.

Sending all together like this is a bit easier to reason about. For example, it's not obvious that `removeNodeUpgrade` in the current API actually marks the node as completed. What happens, for example, if a node is deleted mid-upgrade? How will the status manager detect that and know what to do with its internal caches?

Reply:
The downside is that it's not fine-grained, so you need to call this at the end of every iteration of the reconcile loop rather than as you process each node. However, the reconcile loop never actually exits early (it just skips some nodes), so by the end of the function the `pending`, `inProgress`, and `completed` slices should be an accurate representation of the cluster.

Reply:
Refactored with 80324b4c4c535954326b59a6de69bf8b4cee03e1

I agree the API was a bit weird, I wanted to be consistent with the existing API of `AddXXX`s and `RemoveXXX`s, so the AddWindowsNodeUpgrade was tracking nodes pending and nodes in-progress. Doing the status reporting just once as you suggested makes some of the problems you noted elsewhere in this review go away (e.g. calling AddWindowsNodeUpgrade but not RemoveWindowsNodeUpgrade in some scenarios).